File: src/runtime/config/schemas/deployment.cpp
// Registration for the "deployment" namespace schema.
// Mirrors tensorrt_model_connect/runtime_config/schemas/deployment.py.

#include "trtmc/config/schemas/deployment.h"

#include <any>
#include <set>
#include <string>

namespace trtmc::config::schemas {

namespace {
bool is_valid_provider(const std::any& v) {
    if (v.type() != typeid(std::string))
        return false;
    const auto& provider = std::any_cast<const std::string&>(v);
    return provider == "native_trt" || provider == "tvm_ffi" ||
           provider == "tensorrt-edge-llm";
}

bool is_positive_int(const std::any& v) {
    if (v.type() != typeid(std::int32_t))
        return false;
    return std::any_cast<std::int32_t>(v) > 0;
}
} // namespace

Schema make_deployment_schema() {
    const std::set<Layer> build_and_session = {Layer::BuildTime, Layer::BundleDefault,
                                               Layer::PlatformProfile,
                                               Layer::SessionRequest};
    const std::set<Layer> session = {Layer::PlatformProfile, Layer::SessionRequest};
    return Schema{
        "deployment",
        {
            ConfigField{"target", "string", std::any{std::string{"generic"}},
                        build_and_session, nullptr},
            ConfigField{"provider", "string", std::any{std::string{"native_trt"}},
                        build_and_session, is_valid_provider},
            ConfigField{"variant", "string", std::any{std::string{}}, build_and_session,
                        nullptr},
            ConfigField{"force_fallback", "bool", std::any{false}, session, nullptr},
            ConfigField{"enable_ffi_attention", "bool", std::any{false}, build_and_session,
                        nullptr},
            ConfigField{"ffi_kernel_artifacts", "string", std::any{std::string{}},
                        build_and_session, nullptr},
            ConfigField{"edge_llm_workspace", "string", std::any{std::string{}},
                        build_and_session, nullptr},
            ConfigField{"edge_llm_engine_dir", "string", std::any{std::string{}},
                        build_and_session, nullptr},
            ConfigField{"edge_llm_export_tool", "string",
                        std::any{std::string{"tensorrt-edgellm-export-llm"}},
                        build_and_session, nullptr},
            ConfigField{"edge_llm_build_tool", "string", std::any{std::string{"llm_build"}},
                        build_and_session, nullptr},
            ConfigField{"edge_llm_export_device", "string", std::any{std::string{"cuda"}},
                        build_and_session, nullptr},
            ConfigField{"edge_llm_max_input_len", "int32", std::any{std::int32_t{1024}},
                        build_and_session, is_positive_int},
            ConfigField{"edge_llm_max_batch_size", "int32", std::any{std::int32_t{4}},
                        build_and_session, is_positive_int},
        },
    };
}

REGISTER_CONFIG_SCHEMA_FACTORY_WITH_MANIFEST(register_deployment_schema, make_deployment_schema);
} // namespace trtmc::config::schemas

File: tensorrt_model_connect/tensorrt_model_connect/runtime_config/schemas/deployment.py
"""Schema for the ``deployment`` namespace.

Deployment specialization is selected through the existing two-flag config
surface instead of adding one flag per provider.  The build command consumes
these fields to choose a build provider or kernel specialization; the runtime
command can use the same namespace to force a bundled variant for debugging.
"""

from __future__ import annotations

from tensorrt_model_connect.runtime_config import (
    ConfigField,
    Layer,
    Schema,
    register_schema,
)


_BUILD_AND_SESSION = frozenset({
    Layer.BUILD_TIME,
    Layer.BUNDLE_DEFAULT,
    Layer.PLATFORM_PROFILE,
    Layer.SESSION_REQUEST,
})


SCHEMA = Schema(
    namespace="deployment",
    fields=(
        ConfigField(
            name="target",
            type_tag="string",
            default="generic",
            allowed_layers=_BUILD_AND_SESSION,
        ),
        ConfigField(
            name="provider",
            type_tag="string",
            default="native_trt",
            allowed_layers=_BUILD_AND_SESSION,
            validator=lambda v: v in {
                "native_trt",
                "tvm_ffi",
                "tensorrt-edge-llm",
            },
        ),
        ConfigField(
            name="variant",
            type_tag="string",
            default="",
            allowed_layers=_BUILD_AND_SESSION,
        ),
        ConfigField(
            name="force_fallback",
            type_tag="bool",
            default=False,
            allowed_layers=frozenset({
                Layer.PLATFORM_PROFILE,
                Layer.SESSION_REQUEST,
            }),
        ),
        ConfigField(
            name="enable_ffi_attention",
            type_tag="bool",
            default=False,
            allowed_layers=_BUILD_AND_SESSION,
        ),
        ConfigField(
            name="ffi_kernel_artifacts",
            type_tag="string",
            default="",
            allowed_layers=_BUILD_AND_SESSION,
        ),
        ConfigField(
            name="edge_llm_workspace",
            type_tag="string",
            default="",
            allowed_layers=_BUILD_AND_SESSION,
        ),
        ConfigField(
            name="edge_llm_engine_dir",
            type_tag="string",
            default="",
            allowed_layers=_BUILD_AND_SESSION,
        ),
        ConfigField(
            name="edge_llm_export_tool",
            type_tag="string",
            default="tensorrt-edgellm-export-llm",
            allowed_layers=_BUILD_AND_SESSION,
        ),
        ConfigField(
            name="edge_llm_build_tool",
            type_tag="string",
            default="llm_build",
            allowed_layers=_BUILD_AND_SESSION,
        ),
        ConfigField(
            name="edge_llm_export_device",
            type_tag="string",
            default="cuda",
            allowed_layers=_BUILD_AND_SESSION,
        ),
        ConfigField(
            name="edge_llm_max_input_len",
            type_tag="int32",
            default=1024,
            allowed_layers=_BUILD_AND_SESSION,
            validator=lambda v: isinstance(v, int) and v > 0,
        ),
        ConfigField(
            name="edge_llm_max_batch_size",
            type_tag="int32",
            default=4,
            allowed_layers=_BUILD_AND_SESSION,
            validator=lambda v: isinstance(v, int) and v > 0,
        ),
    ),
)


register_schema(SCHEMA)
